StyleCop.Analyzers/StyleCop.Analyzers/Helpers/DocumentBasedFixAllProvider.cs

```diff
@@ -42,21 +45,26 @@ internal abstract class DocumentBasedFixAllProvider : FixAllProvider
             }
         }
 
-        protected abstract Task<SyntaxNode> FixAllInDocumentAsync(FixAllContext fixAllContext, Document document);
+        protected abstract Task<SyntaxNode> FixAllInDocumentAsync(FixAllContext fixAllContext, Document document, ImmutableArray<Diagnostic> diagnostics);
 
         private async Task<Solution> GetProjectFixesAsync(FixAllContext fixAllContext, Project project)
         {
             Solution solution = project.Solution;
-            var oldDocuments = project.Documents.ToImmutableArray();
-            List<Task<SyntaxNode>> newDocuments = new List<Task<SyntaxNode>>(oldDocuments.Length);
-            foreach (var document in oldDocuments)
+            List<Document> oldDocuments = new List<Document>();
+            List<Task<SyntaxNode>> newDocumentRoots = new List<Task<SyntaxNode>>();
+            var diagnostics = await fixAllContext.GetAllDiagnosticsAsync(project).ConfigureAwait(false);
+            var lookup = diagnostics.GroupBy(x => project.GetDocument(x.Location.SourceTree));
+            foreach (var item in lookup)
             {
-                newDocuments.Add(this.FixAllInDocumentAsync(fixAllContext, document));
+                oldDocuments.Add(item.Key);
+                newDocumentRoots.Add(Task.Run(() => this.FixAllInDocumentAsync(fixAllContext, item.Key, item.ToImmutableArray()), fixAllContext.CancellationToken));
             }
 
-            for (int i = 0; i < oldDocuments.Length; i++)
+            for (int i = 0; i < oldDocuments.Count; i++)
             {
-                var newDocumentRoot = await newDocuments[i].ConfigureAwait(false);
+                fixAllContext.CancellationToken.ThrowIfCancellationRequested();
```

Comment:
:question: Is this actually necessary? It seems like the `await` on the following line would throw if cancellation occurred.